py-horovod: set appropriate build env vars

adamjstewart opened the pull request:
Fixes #15545 

@rblake-llnl @rvinaybharadwaj can you test this? 

P.S. if anyone else would like to be added as a maintainer, let me know.

rblake-llnl commented:
With default options:

```
  >> 125    error: You should not mix NCCL and MPI GPU due to a possible deadlock.
```

When tuning everything to use mpi

```
  >> 138    horovod/common/parameter_manager.h:26:10: fatal error: Eigen/Core: No such file or directory
```

I'll dig into it more in a second.

adamjstewart commented:
@rblake-llnl 

> error: You should not mix NCCL and MPI GPU due to a possible deadlock.

Fixed by setting `HOROVOD_ALLOW_MIXED_GPU_IMPL` and switching defaults.

> fatal error: Eigen/Core: No such file or directory

Fixed by checking out git submodules instead of downloading tarball

@rvinaybharadwaj 

> ValueError: could not convert string to float: 'none'

Fixed by modifying conflict statements. If you use CUDA with PyTorch or TF, you need to also specify the `cuda_arch` of your GPU.

rvinaybharadwaj commented:
> @rblake-llnl
> 
> > error: You should not mix NCCL and MPI GPU due to a possible deadlock.
> 
> Fixed by setting `HOROVOD_ALLOW_MIXED_GPU_IMPL` and switching defaults.
> 
> > fatal error: Eigen/Core: No such file or directory
> 
> Fixed by checking out git submodules instead of downloading tarball
> 
> @rvinaybharadwaj
> 
> > ValueError: could not convert string to float: 'none'
> 
> Fixed by modifying conflict statements. If you use CUDA with PyTorch or TF, you need to also specify the `cuda_arch` of your GPU.

May be disable CUDA build by default

adamjstewart commented:
Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.

rvinaybharadwaj commented:
> Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.

Agreed. However, none of the upcoming exascale machines employ NVIDIA hardware.

rblake-llnl commented:
Works for me, and also runs a hello-world example training mnist

rblake-llnl commented:
> > Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.
> 
> Agreed. However, none of the upcoming exascale machines employ NVIDIA hardware.

I also vote for `+cuda`.  GPUs are so much faster than CPUs that there isn't much point in training without support for them.  People who want to train on exascale should require to take care and take note that they're using the slow version of training.

rvinaybharadwaj commented:
> > > Deep learning without GPU support isn't really useful, I think the default should be `+cuda`.
> > 
> > 
> > Agreed. However, none of the upcoming exascale machines employ NVIDIA hardware.
> 
> I also vote for `+cuda`. GPUs are so much faster than CPUs that there isn't much point in training without support for them. People who want to train on exascale should require to take care and take note that they're using the slow version of training.

My point is different machines use different GPUs and accelerators, and the install should not error out with default options.

adamjstewart commented:
It sounds to me like users of new exascale machines won't even be installing Horovod/TF/PyTorch if that's the case. It's a shame that the people making purchasing decisions aren't aware of the trend towards deep learning in science, but there isn't much we can do about that.

The default variants don't error out, they raise a conflict with a message that explicitly states what they need to do, so there's no confusion. But again, installing a CPU-only version of a deep learning library will result in an order of magnitude performance drop: https://medium.com/@shachishah.ce/do-we-really-need-gpu-for-deep-learning-47042c02efe2. I don't mind the default variants requiring intervention for some users, but I do mind Spack silently installing a version of a deep learning library that's completely useless for 99% of deep learning researchers.

P.S. I'm a deep learning researcher 😄

rvinaybharadwaj commented:
> It sounds to me like users of new exascale machines won't even be installing Horovod/TF/PyTorch if that's the case. It's a shame that the people making purchasing decisions aren't aware of the trend towards deep learning in science, but there isn't much we can do about that.
> 
> The default variants don't error out, they raise a conflict with a message that explicitly states what they need to do, so there's no confusion. But again, installing a CPU-only version of a deep learning library will result in an order of magnitude performance drop: https://medium.com/@shachishah.ce/do-we-really-need-gpu-for-deep-learning-47042c02efe2. I don't mind the default variants requiring intervention for some users, but I do mind Spack silently installing a version of a deep learning library that's completely useless for 99% of deep learning researchers.
> 
> P.S. I'm a deep learning researcher 😄

I don't have any objections if you feel so strongly about it. But the vendors will be supporting different frameworks on their platforms.

darthsuogles commented on var/spack/repos/builtin/packages/py-torch/package.py:
```diff
@@ -105,22 +105,22 @@ class PyTorch(PythonPackage, CudaPackage):
               msg='Must specify CUDA compute capabilities of your GPU, see '
               'https://developer.nvidia.com/cuda-gpus')
-    conflicts('cuda_arch=52', when='@1.3.0:+cuda+caffe2',
+    conflicts('cuda_arch=52', when='@1.3.0:+cuda',
```
What's the reason behind deprecating these versions? 
At least looks like `TORCH_CUDA_ARCH_LIST="3.5 5.2 6.0 6.1 7.0+PTX" ` is still in the official docker image.
https://github.com/pytorch/pytorch/blob/c8f665dcb626378ca9c51d7b70fc242524105441/Dockerfile#L47

adamjstewart replied:
These conflicts were added by @glennpj in #14619, maybe he can explain where they came from. Hopefully I didn't misunderstand the code and the restriction only applies to `+cuda+caffe2`.

adamjstewart replied:
Ping @glennpj. I was able to build PyTorch 1.4.0 with `cuda_arch=35` just fine (although I haven't tested it yet). Where did these conflicts come from?
